File: VRTInterstitialCustomEventIronSource.m
//
//  VRTInterstitialCustomEventIronSource.m
//
//  Created by Scott McCoy on 5/9/19.
//  Copyright © 2019 VRTCAL. All rights reserved.
//

//Header
#import "VRTInterstitialCustomEventIronSource.h"

//Dependencies
#import "VRTIronSourceManager.h"

//IronSource Interstitial Adapter, Vrtcal as Primary
@interface VRTInterstitialCustomEventIronSource() <ISInterstitialDelegate>
@property NSString *instanceId;
@end


@implementation VRTInterstitialCustomEventIronSource

- (void) loadInterstitialAd {
    
    @try {
        
        //Get and validate instance ID
        self.instanceId = [self.customEventConfig.thirdPartyCustomEventData objectForKey:@"instanceId"];
        if (![self stringIsGood:self.instanceId]) {
            VRTError *vrtError = [VRTError errorWithCode:VRTErrorCodeInvalidParam format:@"IronSource instanceId of %@ unusable", self.instanceId];
            [self.customEventLoadDelegate customEventFailedToLoadWithError:vrtError];
            return;
        }
        
        //Get and validate app ID
        NSString *appKey = [self.customEventConfig.thirdPartyCustomEventData objectForKey:@"applicationKey"];
        if (![self stringIsGood:appKey]) {
            VRTError *vrtError = [VRTError errorWithCode:VRTErrorCodeInvalidParam format:@"IronSource appKey of %@ unusable", appKey];
            [self.customEventLoadDelegate customEventFailedToLoadWithError:vrtError];
            return;
        }
        
        //Init IronSource if it hasn't been already
        [[VRTIronSourceManager singleton] initIronSourceSDKWithAppKey:appKey forAdUnits:[NSSet setWithObject:IS_INTERSTITIAL]];
        
        //Check if an interstitial is ready
        if ([IronSource hasInterstitial]) {
            [self.customEventLoadDelegate customEventLoaded];
            return;
        }
        
        //Request an interstitial
        [IronSource setInterstitialDelegate:self];
        [IronSource loadInterstitial];
    } @catch (NSException *exception) {
        
        NSString *description = [exception description];
        VRTError *vrtError = [VRTError errorWithCode:VRTErrorCodeUnknown format:@"Exception in loadInterstitialAd: %@", description];
        [self.customEventLoadDelegate customEventFailedToLoadWithError:vrtError];
    }

}

- (void) showInterstitialAd {
    UIViewController *vc = [self.viewControllerDelegate vrtViewControllerForModalPresentation];
    [IronSource showInterstitialWithViewController:vc];
}



#pragma mark - ISInterstitialDelegate
- (void)didClickInterstitial {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventClicked];
}

- (void)interstitialDidClose {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventDidDismissModal:VRTModalTypeInterstitial];
}

- (void)interstitialDidFailToLoadWithError:(NSError *)error {
    VRTLogWhereAmI();
    [self.customEventLoadDelegate customEventFailedToLoadWithError:error];
}

- (void)interstitialDidFailToShowWithError:(NSError *)error {
    //No VRT analog for this
    VRTLogWhereAmI();
}

- (void)interstitialDidLoad {
    VRTLogWhereAmI();
    [self.customEventLoadDelegate customEventLoaded];
}

- (void)interstitialDidOpen {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventWillPresentModal:VRTModalTypeInterstitial];
}

- (void)interstitialDidShow {
    VRTLogWhereAmI();
    [self.customEventShowDelegate customEventDidPresentModal:VRTModalTypeInterstitial];
}


#pragma mark - Utility
- (BOOL) stringIsGood:(NSString*)str {
    return str != nil && [str isKindOfClass:[NSString class]] && [str length] > 0;
}

@end

File: VRTInterstitialCustomEventIronSource.swift

//
//  VRTInterstitialCustomEventIronSource.swift
//
//  Created by Scott McCoy on 5/9/19.
//  Copyright © 2019 VRTCAL. All rights reserved.
//

//Header
//IronSource Interstitial Adapter, Vrtcal as Primary

class VRTInterstitialCustomEventIronSource: VRTAbstractInterstitialCustomEvent, ISInterstitialDelegate {
    private var instanceId: String?

    func loadInterstitialAd() {

        // TODO: import SwiftTryCatch from https://github.com/ypopovych/SwiftTryCatch
        SwiftTryCatch.try({

            //Get and validate app ID
            let appKey = customEventConfig.thirdPartyCustomEventData["appId"] as? String
            if !stringIsGood(appKey) {
                let vrtError = VRTError(code: VRTErrorCodeInvalidParam, format: "IronSource appId of %@ unusable", appKey)
                customEventLoadDelegate.customEventFailedToLoadWithError(vrtError)
                return
            }

            //Init IronSource if it hasn't been already
            VRTIronSourceManager.singleton().initIronSourceSDK(withAppKey: appKey ?? "", forAdUnits: Set<AnyHashable>([IS_INTERSTITIAL]))

            //Check if an interstitial is ready
            if IronSource.hasInterstitial() {
                customEventLoadDelegate.customEventLoaded()
                return
            }

            //Request an interstitial
            IronSource.interstitialDelegate = self
            IronSource.loadInterstitial()
        }, catch: { exception in

            let description = exception.description()
            let vrtError = VRTError(code: VRTErrorCodeUnknown, format: "Exception in loadInterstitialAd: %@", description)
            customEventLoadDelegate.customEventFailedToLoadWithError(vrtError)
        }, finallyBlock: {
        })

    }

    func showInterstitialAd() {
        let vc = viewControllerDelegate.vrtViewControllerForModalPresentation()
        IronSource.showInterstitial(with: vc)
    }

    // MARK: - ISInterstitialDelegate

    func didClickInterstitial() {
        VRTLogWhereAmI()
        customEventShowDelegate.customEventClicked()
    }

    func interstitialDidClose() {
        VRTLogWhereAmI()
        customEventShowDelegate.customEventDidDismissModal(VRTModalTypeInterstitial)
    }

    func interstitialDidFailToLoadWithError(_ error: Error?) {
        VRTLogWhereAmI()
        customEventLoadDelegate.customEventFailedToLoadWithError(error)
    }

    func interstitialDidFailToShowWithError(_ error: Error?) {
        //No VRT analog for this
        VRTLogWhereAmI()
    }

    func interstitialDidLoad() {
        VRTLogWhereAmI()
        customEventLoadDelegate.customEventLoaded()
    }

    func interstitialDidOpen() {
        VRTLogWhereAmI()
        customEventShowDelegate.customEventWillPresentModal(VRTModalTypeInterstitial)
    }

    func interstitialDidShow() {
        VRTLogWhereAmI()
        customEventShowDelegate.customEventDidPresentModal(VRTModalTypeInterstitial)
    }

    // MARK: - Utility

    func stringIsGood(_ str: String?) -> Bool {
        return str != nil && (str is NSString) && (str?.count ?? 0) > 0
    }
}

//Dependencies

//IronSource Interstitial Adapter, Vrtcal as Primary
